File: ProductofArrayExceptSelf.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int leftptr = 1; 
        int rightptr = 1; 
        int[] ans = new int[nums.length];
        int[] left = new int[nums.length]; 
        int[] right = new int[nums.length];
        for (int i = 0; i < nums.length; i++){
            left[i] = leftptr;
            leftptr *= nums[i];
        }
        for (int i = nums.length - 1; i >= 0; i--){
            right[i] = rightptr;
            rightptr *= nums[i];
        }
        for (int i = 0; i < nums.length; i++){
            ans[i] = left[i] * right[i];
        }
        return ans;
    }
}

File: ProductOfArrayExceptSelf.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left, right = [1] * (len(nums) + 1), [1]* (len(nums) + 1)
        #build the left
        for i in range(len(nums)):
            left[i] *= nums[i] * left[i - 1]

        #build the right
        for i in range(len(nums) - 1, -1, -1):
            right[i] *= nums[i] * right[i+1]

        #product here
        for i in range(len(nums)):
            nums[i] = left[i - 1] * right[i + 1]
        
        return nums
            
